timestats/transform.go

```diff
@@ -0,0 +1,46 @@
+package timestats
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/montanaflynn/stats"
+)
+
+// Floater is the interface to get a float64 value from a struct.
+type Floater interface {
+	Float() float64
+}
+
+// TransformIter returns a data set ready for computation.
+// The data set is built with floats retrieved from x.
+// x must be a slice of structs implementing timestats.Floater.
+func TransformIter(x interface{}) (stats.Float64Data, error) {
+	s := reflect.ValueOf(x)
+	if s.Kind() != reflect.Slice {
+		return nil, fmt.Errorf("received a non-slice type: %s", reflect.TypeOf(x))
+	}
+
+	floats := make(stats.Float64Data, s.Len())
+
+	for i := 0; i < s.Len(); i++ {
+		v := s.Index(i)
+		f, ok := v.Interface().(Floater)
+		if !ok {
+			return nil, fmt.Errorf("%s %w", v.Type(), ErrInterface)
+		}
+		floats[i] = f.Float()
+	}
+	return floats, nil
+}
+
+// FloatSlicer is the interface to get a slice of float64 from a type.
+type FloatSlicer interface {
+	FloatSlice() []float64
+}
+
+// Transform returns a data set ready for computation.
+// The data set is extracted from the given FloatSlicer.
+func Transform(in FloatSlicer) stats.Float64Data {
+	return in.FloatSlice()
+}
```

Comment:
[question] I'm wondering whether `FloatSlicer` is really relevant 🤔  I mean if we expect the caller to know how to convert its input to `[]float64`, this is equivalent to:

```go
// timestats/transform.go
func Cast(in []float64) stats.Float64Data { return in }

// main.go
func main() {
	fs := myFloatSlicer{}
	floatData := timestats.Cast(fs.FloatSlice())
}
```

And at this point, couldn't `timestats` just accept `[]float64` and make the necessary casting internally just using `stats.Float64Data(input)`? So we would just do in main:

```go
func main() {
	report := readReport()
	computed, err := timestats.Compute(report.Floats())
}
```

Reply:
Well yes it could ! It need to think a bit more about it, for now it find it elegant that whatever happens with `timestats` internally, `Transform` serves as a clearer boundary timestats-side. It used to do validations and return an error.
I'd say we can come back and simplify later and keep it as is for now. But if its a real problem for you I'll change right now

Reply:
Well the computing process seems a bit weird for now:

1. We start with a`[]time.Duration` input (not exactly but anyways)
2. We use `timestats` to handle stats with durations, but it actually doesn't because it requires `[]float64`, so we handle the conversion `[]time.Duration` ➡️ `[]float64`  in `main` anyway
3. We use `timestats.Transform` that accepts an interface but just calls it and returns its result casted to `stats.Float64Data`
4. We finally call `timestats.Compute` with the gotten `stats.Float64Data`

Point 2: we agreed we can iterate later 👍 

What bothers me with 3 and 4  is without any change in `timestats`, we can already just pass `[]float64` to `timestats.Compute`, because `stats.Float64Data` is `[]float64`. Two consequences:
  1. we really don't need `Transform`, even in current state, so we can delete it (and the interface)
  1. we can change `Compute` param signature to `func([]float64)` without any further adaptation and it would work equally: that way we even avoid to expose types from `stats` which is kind of the point of having a package `timestats`

Given how small and easy this change is I see no reason not to do it here, but we can't really afford to block the merge for these considerations so I'll approve and leave you with the decision.

Reply:
Well you're right.
`timestats` started as a package capable of dealing with any struct exposing a `time.Duration` field. But it was a bit tedious (generics would have help, maybe ?) and I went the "you must implement my interface" way.

But as you painted it, I agree it's suoerflous. I'll cut everything :ok_hand:

Reply:
Yeah that genericity was not achievable without being either over-engineered or on the contrary superfluous, and I'm not even convinced generics would have been of any help (maybe genericity is just not what we need here 😕). But anyway, it was worth experimenting!